packages/animations/src/version.ts

```diff
@@ -24,9 +24,9 @@ export class Version {
 
   constructor(public full: string) {
     const splits = full.split('.');
-    this.major = splits[0];
-    this.minor = splits[1];
-    this.patch = splits.slice(2).join('.');
+    this.major = splits.shift()!;
+    this.minor = splits.shift()!;
+    this.patch = splits.join('.');
```

Comment:
May be we can refactor the code like this?
```suggestion
    const [major, minor, ...rest] = full.split('.');
    this.major = major;
    this.minor = minor;
    this.patch = rest.join('.');
```

Reply:
This is a bit uglier but less bytes:
```ts
    let rest: string[];
    [this.major, this.minor, ...rest] = full.split('.');
    this.patch = rest.join('.');
```

Or if RegEx is not a problem:
```ts
    [, this.major, this.minor, this.patch] = full.match(/^(\d+)\.(\d+)\.(.+)/)!;
```

Reply:
I also thought about using array destructuring, but I was afraid it could be slower, so I made a little benchmark :

<details>
  <summary>Code used for the benchmark</summary>

  Tested on an old `i7 4712HQ` cpu

  ```typescript
  class Version {
    public major: string;
    public minor: string;
    public patch: string;

    constructor(public full: string) {
      const numberOfTests = 100000000;

      console.time('split');
      for (let i = 0; i < numberOfTests; i++) {
        this.versionSplit(full);
      }
      console.timeEnd('split');

      console.time('destructuring split');
      for (let i = 0; i < numberOfTests; i++) {
        this.versionDestructuringSplit(full);
      }
      console.timeEnd('destructuring split');

      console.time('condensed destructuring split');
      for (let i = 0; i < numberOfTests; i++) {
        this.versionCondensedDestructuringSplit(full);
      }
      console.timeEnd('condensed destructuring split');

      console.time('regex');
      for (let i = 0; i < numberOfTests; i++) {
        this.versionRegex(full);
      }
      console.timeEnd('regex'),

      console.time('condensed regex');
      for (let i = 0; i < numberOfTests; i++) {
        this.versionCondensedRegex(full);
      }
      console.timeEnd('condensed regex');
    }

    versionSplit(full: string) {
      const splits = full.split('.');
      this.major = splits.shift()!;
      this.minor = splits.shift()!;
      this.patch = splits.join('.');
    }

    versionDestructuringSplit(full: string) {
      const [major, minor, ...rest] = full.split('.');
      this.major = major;
      this.minor = minor;
      this.patch = rest.join('.');
    }

    versionCondensedDestructuringSplit(full: string) {
      let rest: string[];
      [this.major, this.minor, ...rest] = full.split('.');
      this.patch = rest.join('.');
    }

    versionRegex(full: string) {
      const match = full.match(/^(\d+)\.(\d+)\.(.+)/)!;
      this.major = match[1];
      this.minor = match[2];
      this.patch = match[3];
    }

    versionCondensedRegex(full: string) {
      [, this.major, this.minor, this.patch] = full.match(/^(\d+)\.(\d+)\.(.+)/)!;
    }
  }

  new Version('12.0.0');
  new Version('12.0.0-next.1');
  ```
</details>

I was not expecting the RegEx solution to be the fastest ^^

Here are the results I got using Node.js v15 (Node.js v12 gave similar results) :

|                                 | `12.0.0` + ES6 | `12.0.0-next.1` + ES6 | `12.0.0` + ES5 | `12.0.0-next.1` + ES5 |
| :------------------------------ | -------------: | --------------------: | -------------: | --------------------: |
| split                           |          26.2s |                 32.9s |          30.7s |                 33.8s |
| split + destructuring           |          27.9s |                 35.3s |          41.0s |                 48.9s | 
| split + condensed destructuring |          27.5s |                 34.2s |          41.5s |                 47.8s |
| regex                           |          11.4s |                 13.2s |          12.1s |                 13.0s |
| condensed regex                 |          11.8s |                 13.6s |          25.3s |                 25.9s |

For completness, I also tested this using Firefox and altough Firefox got better result  than Node.js for `split + "12.0.0" + ES6` (17.5s instead of 26.2s), the `regex` solution was still better (12.5s)

Transpiling to ES5 basically transformed the rest operator with a `slice()` call, which I wanted to avoid in the first place by using `shift()`, and as expected, using `slice()` yielded the worst results

Array destructuring suffers from a big performance penalty in ES5, since its wrapping the line with a `__read()` call (which is why the `condensed regex` solution is 2 times slower with ES5 than ES6)

As I expect more and more people to use ES6 or newer versions of EcmaScript, the `condensed regex` solution appears to me as the best compromise between code size & performance, what do you guys think ?

Reply:
Code/size performance, yes I would have to agree

Reply:
Hi @Julien-Marcou, thanks for measuring performance.

Since the code that you've updated is not on a critical path (i.e. it's not being executed frequently) and in fact it's executed just once per page load to create an instance of the `Version` class, I'd vote for more simple/readable implementation that would simplify future maintenance.

Reply:
I agree, I have no preference between a RegEx or using `split`, but a RegEx is indeed harder to maintain.

Should I go for the `split + destructuring` then ? @AndrewKushnir

Reply:
@Julien-Marcou I'd vote for `split + destructuring` approach 👍

Reply:
To mix things up, I'd personally vote for the nice, declarative regex (but I am known for being biased towards regexes, so that probably doesn't count 😛).